torch/csrc/autograd/functions/tensor.cpp

```diff
@@ -17,24 +17,27 @@ namespace torch { namespace autograd {
 
 auto CopyBackwards::apply(variable_list&& grads) -> variable_list {
   check_input_variables("CopyBackwards", grads, 1, -1, true);
-  auto& grad = grads[0];
+  auto grad = c10::MaybeOwned<at::Tensor>::borrowed(grads[0]);
   variable_list grad_inputs(2);
-  if (grad.defined()) {
+  if (grad->defined()) {
     if (should_compute_output(0)) {
-      grad_inputs[0] = at::zeros_like(grad, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+      grad_inputs[0] = at::zeros_like(*grad, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
     }
     if (should_compute_output(1)) {
+      // Handle R->C copies without raising a warning
+      const auto src_type = src_options.dtype().toScalarType();
+      if (!c10::isComplexType(src_type) && grad->is_complex()) {
+        grad = c10::MaybeOwned<at::Tensor>::owned(at::real(grads[0]));
+      }
+
       at::DeviceGuard device_guard(src_device);
       // TODO: What if !grad.is_cuda(), but src_device is CUDA?
       // This code is kind of weirdly asymmetric.
-      if (grad.is_cuda() && grad.device() != src_device) {
-        grad_inputs[1] = grad.to(
-            src_options,
-            /*non_blocking=*/false,
-            /*copy=*/true);
-      } else {
-        grad_inputs[1] = grad.to(src_options);
-      }
+      const bool copy = (grad->is_cuda() && grad->device() != src_device);
+      grad_inputs[1] = grad->to(
+          src_options,
+          /*non_blocking=*/false,
+          /*copy=*/copy);
```

Comment:
Why is this used? If grad is not on the same device as src, the `to()` here is going to trigger a copy no?

Reply:
I agree, opened #60025 to clean this code up.

Reply:
Thanks!